### PR TITLE
Show vulnerabilities WebApp report

### LINKED ISSUE #4630: Make the WebApp report list vulnerabilities
Currently, only the [evaluated model](https://github.com/oss-review-toolkit/ort/pull/4306) and vulnerability reports list vulnerabilities, but [users rightfully assume](https://github.com/oss-review-toolkit/ort/issues/4219#issuecomment-953559961) the WebApp report to also list vulnerabilities.

### PR DESCRIPTION
**Pull request by tsteenbe**
Initial implementation to show vulnerabilities in the WebApp report.

Closes #4630

### CODE REVIEW

File: reporter-web-app/src/components/VulnerabilitiesTable.js [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+            showExcludesColumn,
+            state
+        } = this.props;
+        const {
+            filteredInfo = {},
+            sortedInfo = {}
+        } = state;
+
+        const renderAhref = (element, href, key, target = '_blank') => (
+            <a
```
[the remark this answers is not in the document]
Comment (tsteenbe) [reply reporter-web-app/src/components/VulnerabilitiesTable.js:60]: @sonatype-lift ignore

File: reporter-web-app/src/models/VulnerabilityReference.js [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+class VulnerabilityReference {
+    #scoringSystem;
```
[the remark this answers is not in the document]
Comment (tsteenbe) [reply reporter-web-app/src/models/VulnerabilityReference.js:21]: @sonatype-lift ignore

File: reporter-web-app/src/models/WebAppVulnerability.js [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+import { randomStringGenerator } from '../utils';
+import VulnerabilityReference  from './VulnerabilityReference';
+
+class WebAppVulnerability {
+    #_id;
```
[the remark this answers is not in the document]
Comment (tsteenbe) [reply reporter-web-app/src/models/WebAppVulnerability.js:24]: @sonatype-lift ignore

File: reporter-web-app/src/models/VulnerabilityReference.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            }
+
+            if (obj.severity !== null) {
+                this.#severity = obj.severity;
+            }
+
+            if (obj.severity_mapped !== null || obj.severityMapped !== null) {
+                this.#severityMapped = obj.severity_mapped || obj.severityMapped;
+
+                switch (this.#severityMapped) {
```
[the remark this answers is not in the document]
Comment (tsteenbe) [reply reporter-web-app/src/models/VulnerabilityReference.js:44]: @sonatype-lift ignore

File: reporter-web-app/src/models/VulnerabilityReference.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+                    case 'CRITICAL':
+                        this.#severityIndex = 0;
+                        break;
+                    case 'HIGH':
```
[the remark this answers is not in the document]
Comment (tsteenbe) [reply reporter-web-app/src/models/VulnerabilityReference.js:48]: @sonatype-lift ignore

[the remark this answers is not in the document]
Comment (tsteenbe) [reply reporter-web-app/src/models/VulnerabilityReference.js:48]: @sonatype-lift ignore

File: reporter-web-app/src/models/VulnerabilityReference.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (tsteenbe) [reply reporter-web-app/src/models/VulnerabilityReference.js:44]: @sonatype-lift ignore

File: reporter-web-app/src/models/WebAppVulnerability.js [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+            }
+
+            if (obj.resolutions) {
+                this.#resolutionIndexes = new Set(obj.resolutions);
+            }
+
+            if (webAppOrtResult) {
+                this.#webAppOrtResult = webAppOrtResult;
+
+                const webAppPackage = webAppOrtResult.getPackageByIndex(this.#packageIndex);
```
[the remark this answers is not in the document]
Comment (tsteenbe) [reply reporter-web-app/src/models/WebAppVulnerability.js:74]: @sonatype-lift ignore

File: reporter-web-app/src/models/VulnerabilityReference.js [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+                    case 'LOW':
+                        this.#severityIndex = 3;
+                        break;
+                    default:
+                        this.#severityIndex = 4;
+                }
+            }
+
+            if (obj.url !== null) {
+                this.#url= obj.url;
```
[the remark this answers is not in the document]
Comment (tsteenbe) [reply reporter-web-app/src/models/VulnerabilityReference.js:63]: @sonatype-lift ignore

File: reporter-web-app/src/models/VulnerabilityReference.js [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+                        break;
+                    case 'LOW':
+                        this.#severityIndex = 3;
+                        break;
+                    default:
+                        this.#severityIndex = 4;
+                }
+            }
+
+            if (obj.url !== null) {
```
[the remark this answers is not in the document]
Comment (tsteenbe) [reply reporter-web-app/src/models/VulnerabilityReference.js:62]: @sonatype-lift ignore

File: reporter-web-app/src/models/WebAppVulnerability.js [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    get _id() {
+        return this.#_id;
+    }
+
+    get id() {
+        return this.#id;
+    }
+
+    get isResolved() {
+        if (this.#resolutionIndexes && this.#resolutionIndexes.size > 0) {
```
[the remark this answers is not in the document]
Comment (tsteenbe) [reply reporter-web-app/src/models/WebAppVulnerability.js:93]: @sonatype-lift ignore

File: reporter-web-app/src/models/WebAppVulnerability.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (tsteenbe) [reply reporter-web-app/src/models/WebAppVulnerability.js:74]: @sonatype-lift ignore

File: reporter-web-app/src/models/WebAppVulnerability.js [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+            if (obj.resolutions) {
+                this.#resolutionIndexes = new Set(obj.resolutions);
+            }
+
+            if (webAppOrtResult) {
+                this.#webAppOrtResult = webAppOrtResult;
+
+                const webAppPackage = webAppOrtResult.getPackageByIndex(this.#packageIndex);
+                if (webAppPackage) {
+                    this.#package = webAppPackage;
```
[the remark this answers is not in the document]
Comment (tsteenbe) [reply reporter-web-app/src/models/WebAppVulnerability.js:76]: @sonatype-lift ignore

File: reporter-web-app/src/models/VulnerabilityReference.js [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+                this.#url= obj.url;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (tsteenbe) [reply reporter-web-app/src/models/VulnerabilityReference.js:65]: @sonatype-lift ignore

File: reporter-web-app/src/models/VulnerabilityReference.js [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+            if (obj.severity_mapped !== null || obj.severityMapped !== null) {
+                this.#severityMapped = obj.severity_mapped || obj.severityMapped;
+
+                switch (this.#severityMapped) {
+                    case 'CRITICAL':
+                        this.#severityIndex = 0;
+                        break;
+                    case 'HIGH':
+                        this.#severityIndex = 1;
```
[the remark this answers is not in the document]
Comment (tsteenbe) [reply reporter-web-app/src/models/VulnerabilityReference.js:49]: @sonatype-lift ignore

File: reporter-web-app/src/models/VulnerabilityReference.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (tsteenbe) [reply reporter-web-app/src/models/VulnerabilityReference.js:21]: @sonatype-lift ignore

File: reporter-web-app/src/models/VulnerabilityReference.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (tsteenbe) [reply reporter-web-app/src/models/VulnerabilityReference.js:63]: @sonatype-lift ignore

File: reporter-web-app/src/models/WebAppVulnerability.js [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+
+    #package;
+
+    #packageIndex;
+
+    #references = [];
+
+    #resolutionIndexes = new Set();
+
+    #resolutionReasons;
```
[the remark this answers is not in the document]
Comment (tsteenbe) [reply reporter-web-app/src/models/WebAppVulnerability.js:36]: @sonatype-lift ignore

File: reporter-web-app/src/models/WebAppVulnerability.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (tsteenbe) [reply reporter-web-app/src/models/WebAppVulnerability.js:93]: @sonatype-lift ignore

File: reporter-web-app/src/models/WebAppVulnerability.js [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+
+    #id;
+
+    #package;
+
+    #packageIndex;
+
+    #references = [];
+
+    #resolutionIndexes = new Set();
```
[the remark this answers is not in the document]
Comment (tsteenbe) [reply reporter-web-app/src/models/WebAppVulnerability.js:34]: @sonatype-lift ignore

File: reporter-web-app/src/models/WebAppVulnerability.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (tsteenbe) [reply reporter-web-app/src/models/WebAppVulnerability.js:93]: @sonatype-lift ignore

File: reporter-web-app/src/models/VulnerabilityReference.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (tsteenbe) [reply reporter-web-app/src/models/VulnerabilityReference.js:21]: @sonatype-lift ignore

File: reporter-web-app/src/models/VulnerabilityReference.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (tsteenbe) [reply reporter-web-app/src/models/VulnerabilityReference.js:44]: @sonatype-lift ignore

File: reporter-web-app/src/models/WebAppVulnerability.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (tsteenbe) [reply reporter-web-app/src/models/WebAppVulnerability.js:93]: @sonatype-lift ignore

File: reporter-web-app/src/components/VulnerabilitiesTable.js [hunk 15]
```diff
@@ -0,0 +1,383 @@
...
+                children={element}
+                href={href}
+                key={key}
+                rel="noopener noreferrer"
+                target={target}
+            >
+            </a>
```
[the remark this answers is not in the document]
Comment (tsteenbe) [reply reporter-web-app/src/components/VulnerabilitiesTable.js:67]: @sonatype-lift ignore

File: reporter/src/funTest/assets/evaluated-model-reporter-test-expected-output.json [hunk 16]
```diff
@@ -833,12 +833,14 @@
   } ],
   "vulnerabilities_resolutions" : [ ],
   "vulnerabilities" : [ {
+    "_id" : 0,
```
Comment (mnonnenmacher) [inline reporter/src/funTest/assets/evaluated-model-reporter-test-expected-output.json:836]: This commit should be squashed with the previous commit following the principle that every single commit of a PR should pass the tests.

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

**Comment by sschuberth** [comment]
Could we maybe get a screenshot of how this looks like in the current state?

File: reporter/src/funTest/assets/evaluated-model-reporter-test-expected-output.json (hunk 16 above)
Comment (tsteenbe) [reply reporter/src/funTest/assets/evaluated-model-reporter-test-expected-output.json:836]: Squashed the commit

**Comment by mnonnenmacher** [comment]
> Could we maybe get a screenshot of how this looks like in the current state?

There is one on Slack: https://ort-talk.slack.com/archives/C9NNJ54B1/p1637677784105200?thread_ts=1637668788.095300&cid=C9NNJ54B1

**Comment by sschuberth** [comment]
> > Could we maybe get a screenshot of how this looks like in the current state?
> 
> There is one on Slack: https://ort-talk.slack.com/archives/C9NNJ54B1/p1637677784105200?thread_ts=1637668788.095300&cid=C9NNJ54B1

Right, but that before adding the severities, or?

**Comment by mnonnenmacher** [comment]
> > > Could we maybe get a screenshot of how this looks like in the current state?
> > 
> > 
> > There is one on Slack: https://ort-talk.slack.com/archives/C9NNJ54B1/p1637677784105200?thread_ts=1637668788.095300&cid=C9NNJ54B1
> 
> Right, but that before adding the severities, or?

They are encoded in the icons, see: https://ort-talk.slack.com/archives/C9NNJ54B1/p1638215346139200?thread_ts=1638180931.138600&cid=C9NNJ54B1

**Comment by sschuberth** [comment]
> They are encoded in the icons, see: https://ort-talk.slack.com/archives/C9NNJ54B1/p1638215346139200?thread_ts=1638180931.138600&cid=C9NNJ54B1

Ah, somehow that thread didn't update for me in Slack, and I didn't see @tsteenbe's response. Thanks for pointing this out!

**Comment by sschuberth** [comment]
Still pasting the screenshot also here:

image

**Comment by sschuberth** [comment]
@tsteenbe could you please rebase once more?

**Comment by sschuberth** [comment]
Ignoring the unrelated failing tests.

**Review by sschuberth — APPROVED** [review]

**Pull request merged**